plugin/test/Types.hs

```diff
@@ -0,0 +1,127 @@
+{-# LANGUAGE FlexibleInstances     #-}
+{-# LANGUAGE GADTs                 #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+
+module Types where
+
+import qualified ConCat.Category as CC
+import           Data.Typeable
+import           Unsafe.Coerce (unsafeCoerce)
+
+
+----------------------------------------------------------------------------
+-- | The free category construction over all of the relevant concat classes.
+data FreeSyn a b where
+  CId    :: FreeSyn a b
+  CComp  :: FreeSyn b c -> FreeSyn a b -> FreeSyn a c
+  CTerm  :: FreeSyn a ()
+  CExl   :: FreeSyn (a, b) a
+  CExr   :: FreeSyn (a, b) b
+  CPAnd  :: FreeSyn a b -> FreeSyn a c -> FreeSyn a (b, c)
+  CCurry :: FreeSyn (a, b) c -> FreeSyn a (b -> c)
+  CApply :: FreeSyn (a -> b, a) b
+  CMul   :: FreeSyn (a, a) a
+  CAdd   :: FreeSyn (a, a) a
+  CPow   :: FreeSyn (a, Int) a
+  CNeg   :: FreeSyn a a
+  CInl   :: FreeSyn a (Either a b)
+  CInr   :: FreeSyn b (Either a b)
+  CCOr   :: FreeSyn a c -> FreeSyn b c -> FreeSyn (Either a b) c
+  CConst :: (Show b, Eq b, Typeable b) => b -> FreeSyn a b
+
+
+instance Show (FreeSyn a b) where
+  show CId         = "id"
+  show (CComp a b) = "(comp " ++ show a ++ " " ++ show b ++ ")"
+  show CTerm       = "unit"
+  show CExl        = "exl"
+  show CExr        = "exr"
+  show (CPAnd a b) = "(and " ++ show a ++ " " ++ show b ++ ")"
+  show (CCurry a)  = "(curry " ++ show a ++ ")"
+  show CApply      = "app"
+  show CMul        = "mul"
+  show CAdd        = "add"
+  show CPow        = "pow"
+  show CNeg        = "neg"
+  show CInl        = "inl"
+  show CInr        = "inr"
+  show (CCOr a b)  = "(or " ++ show a ++ " " ++ show b ++ ")"
+  show (CConst a)  = "(const " ++ show a ++ ")"
+
+
+instance Eq (FreeSyn a b) where
+  CId == CId = True
+  CComp a b == CComp c d =
```

Comment:
You could probably define `Ok FreeSyn` to eliminate the need for `unsafeCoerce`.